# Proyek Analisis Data: E-commerce-public-dataset
- **Nama:** Halida Fiadnin
- **Email:** halidaafiadnin@gmail.com
- **ID Dicoding:** halidafi

## Menentukan Pertanyaan Bisnis

1. Bagaimana pola pembelian pelanggan berdasarkan waktu?
2. Apakah ada kategori produk tertentu yang paling sering dibeli?
3. Apakah ada hubungan antara tren pembelian dan lokasi?

## Import Semua Packages/Library yang Digunakan

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [2]:
#orders_dataset
orders_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\orders_dataset.csv")

#order_items_dataset
order_items_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\order_items_dataset.csv")

#products_dataset
products_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\products_dataset.csv")

#product_category_name_translation
product_category_name_translation_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\product_category_name_translation.csv")

#customers_dataset
customers_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\customers_dataset.csv")

#geolocation_dataset
geolocation_df = pd.read_csv(r"C:\Users\halid\Laskar-AI\submission-belajar-analis-data\data\geolocation_dataset.csv")

### Assesing Data

#### orders_dataset

In [13]:
orders_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [14]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [15]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [16]:
print("jumlah duplikasi data: ", orders_df.duplicated().sum())

jumlah duplikasi data:  0


**Insight:**

- orders_dataset **memiliki missing value** pada beberapa atribut seperti order_approved_at, order_delivered_carrier_date, dan order_delivered_customer_date.
- orders_dataset **tidak memiliki duplikasi data**.

#### order_items_dataset

In [17]:
order_items_df.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [18]:
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [19]:
order_items_df.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [20]:
print("jumlah duplikasi data: ", order_items_df.duplicated().sum())

jumlah duplikasi data:  0


**Insight:**

- order_items_dataset **tidak memiliki missing value**
- order_items_dataset **tidak memiliki duplikasi data**.

#### products_dataset

In [21]:
products_df.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [22]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [23]:
products_df.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [24]:
print("jumlah duplikasi data: ", products_df.duplicated().sum())

jumlah duplikasi data:  0


**Insight:**

- products_dataset **memiliki missing value** pada beberapa atribut seperti product_category_name, product_category_lenght, dan lainnya.
- products_dataset **tidak memiliki duplikasi data**.

#### product_category_name_translation

In [25]:
product_category_name_translation_df.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [26]:
product_category_name_translation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [27]:
product_category_name_translation_df.isna().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [28]:
print("jumlah duplikasi data: ", product_category_name_translation_df.duplicated().sum())

jumlah duplikasi data:  0


**Insight:**

- products_category_name_translation **tidak memiliki missing value**.
- products_category_name_translation **tidak memiliki duplikasi data**.

#### customers_dataset

In [29]:
customers_df.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [30]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [31]:
customers_df.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [32]:
print("jumlah duplikasi data: ", customers_df.duplicated().sum())

jumlah duplikasi data:  0


**Insight:**

- customers_dataset **tidak memiliki missing value**.
- customers_dataset **tidak memiliki duplikasi data**.

#### geolocation_dataset

In [33]:
geolocation_df.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [34]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [35]:
geolocation_df.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [36]:
print("jumlah duplikasi data: ", geolocation_df.duplicated().sum())

jumlah duplikasi data:  261831


**Insight:**

- geolocation_dataset **tidak memiliki missing value**.
- geolocation_dataset **terdapat 261831 data yang duplikasi**.

#### Kesimpulan dari assesing data

In [37]:
data = {
    "dataset": [
        "orders_df", "order_items_df", 
        "products_df", "product_category_name_translation_df",
       "customers_df", "geolocation_dataset"
    ],
    "isna": ["ya", "tidak", "ya", "tidak", "tidak", "tidak"],
    "duplikat": ["tidak", "tidak", "tidak", "tidak", "tidak", "ya"]
}

df = pd.DataFrame(data)

df

,dataset,isna,duplikat
0,orders_df,ya,tidak
1,order_items_df,tidak,tidak
2,products_df,ya,tidak
3,product_category_name_translation_df,tidak,tidak
4,customers_df,tidak,tidak
5,geolocation_dataset,tidak,ya


**Insight:**
- terdapat missing value pada dataset **orders_df** dan **products_df**.
- tidak terdapat duplikasi data pada dataset manapun.

### Cleaning Data

#### Cleaning data pada orders_df

In [4]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [5]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**penjelasan:**

pada orders_df, terdapat 3 atribut yang memiliki missing value. untuk menanganinya, 
- **order_approved_at** bisa diisi dengan waktu saat pembelian (order_purchase_timestamp) sebagai perkiraan
- **order_delivered_carrier_date** bisa diisi dengan tanggal perkiraan rata-rata setelah approved
- **order_delivered_customer_date** bisa diisi dengan estimasi waktu pengiriman rata-rata

sebenarnya, order_delivered_carrier_date yang kosong bisa jadi disebabkan karena pesanan tidak pernah dikirim ke kurir.
atau dengan kata lain, bukanlah "delivered" tetapi "canceled" atau "processing".
begitu juga dengan order_delivered_customer_date yang kosong bisa jadi disebabkan karena pesanan belum dikirim, atau pesanan dibatalkan.

In [6]:
# mengisi missing value order_approved_at
orders_df["order_approved_at"].fillna(orders_df["order_purchase_timestamp"], inplace=True)

C:\Users\halid\AppData\Local\Temp\ipykernel_10884\722979261.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  orders_df["order_approved_at"].fillna(orders_df["order_purchase_timestamp"], inplace=True)


**penjelasan:**
- untuk melakukan perhitungan tanggal pengiriman rata-rata dan waktu pengiriman rata-rata, atribut perlu diubah tipe datanya menjadi "datetime"

In [7]:
orders_df["order_approved_at"] = pd.to_datetime(orders_df["order_approved_at"])
orders_df["order_delivered_carrier_date"] = pd.to_datetime(orders_df["order_delivered_carrier_date"])
orders_df["order_delivered_customer_date"] = pd.to_datetime(orders_df["order_delivered_customer_date"])

In [8]:
# mengisi missing value pada order_delivered_carrier_date
avg_shipping_time = (orders_df["order_delivered_carrier_date"] - orders_df["order_approved_at"]).median()
orders_df["order_delivered_carrier_date"].fillna(orders_df["order_approved_at"] + avg_shipping_time, inplace=True)

C:\Users\halid\AppData\Local\Temp\ipykernel_10884\869850625.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  orders_df["order_delivered_carrier_date"].fillna(orders_df["order_approved_at"] + avg_shipping_time, inplace=True)


In [9]:
# mengisi missing value pada order_delivered_customer_date
avg_delivery_time = (orders_df["order_delivered_customer_date"] - orders_df["order_delivered_carrier_date"]).median()
orders_df["order_delivered_customer_date"].fillna(orders_df["order_delivered_carrier_date"] + avg_delivery_time, inplace=True)

C:\Users\halid\AppData\Local\Temp\ipykernel_10884\2590861950.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  orders_df["order_delivered_customer_date"].fillna(orders_df["order_delivered_carrier_date"] + avg_delivery_time, inplace=True)


In [10]:
orders_df.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [11]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  object        
 4   order_approved_at              99441 non-null  datetime64[ns]
 5   order_delivered_carrier_date   99441 non-null  datetime64[ns]
 6   order_delivered_customer_date  99441 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  object        
dtypes: datetime64[ns](3), object(5)
memory usage: 6.1+ MB


**Insight:**
- pengisian missing value pada 3 atribut berhasil dilakukan.
- terdapat perubahan tipe data dari object menjadi datetime pada 3 atribut tersebut.

#### cleaning data pada products_df

In [12]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [13]:
products_df.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

**penjelasan:**

pada orders_df, terdapat **8 atribut** yang memiliki missing value. untuk menanganinya, 
- **product_category_name** dapat diisi dengan "unknown"
- untuk **7 atribut lainnya** dapat diisi dengan median / rata-ratanya.

secara umum, untuk data kuantitatif kita dapat menggantinya dengan nilai median atau rata-rata supaya tidak terpengaruh oleh outlier.

In [14]:
# menangani missing value product_category_name
products_df["product_category_name"].fillna("unknown", inplace=True)

C:\Users\halid\AppData\Local\Temp\ipykernel_10884\1511206838.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  products_df["product_category_name"].fillna("unknown", inplace=True)


In [15]:
# menangan missing value data kuantitatif
products_df["product_name_lenght"].fillna(products_df["product_name_lenght"].median(), inplace=True)
products_df["product_description_lenght"].fillna(products_df["product_description_lenght"].median(), inplace=True)
products_df["product_photos_qty"].fillna(products_df["product_photos_qty"].median(), inplace=True)
products_df["product_weight_g"].fillna(products_df["product_weight_g"].median(), inplace=True)
products_df["product_length_cm"].fillna(products_df["product_length_cm"].median(), inplace=True)
products_df["product_height_cm"].fillna(products_df["product_height_cm"].median(), inplace=True)
products_df["product_width_cm"].fillna(products_df["product_width_cm"].median(), inplace=True)

C:\Users\halid\AppData\Local\Temp\ipykernel_10884\2877992896.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  products_df["product_name_lenght"].fillna(products_df["product_name_lenght"].median(), inplace=True)
C:\Users\halid\AppData\Local\Temp\ipykernel_10884\2877992896.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we a

In [16]:
products_df.isna().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [17]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32951 non-null  object 
 2   product_name_lenght         32951 non-null  float64
 3   product_description_lenght  32951 non-null  float64
 4   product_photos_qty          32951 non-null  float64
 5   product_weight_g            32951 non-null  float64
 6   product_length_cm           32951 non-null  float64
 7   product_height_cm           32951 non-null  float64
 8   product_width_cm            32951 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


**Insight:**
- pengisian missing value pada 8 atribut berhasil dilakukan.

#### cleaning data pada geolocation_df

In [18]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [19]:
print("jumlah duplikasi: ", geolocation_df.duplicated().sum())

jumlah duplikasi:  261831


In [20]:
# menangani nilai duplikat
geolocation_df.drop_duplicates(inplace=True)

In [21]:
print("jumlah duplikasi: ", geolocation_df.duplicated().sum())

jumlah duplikasi:  0


In [22]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 738332 entries, 0 to 1000161
Data columns (total 5 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   geolocation_zip_code_prefix  738332 non-null  int64  
 1   geolocation_lat              738332 non-null  float64
 2   geolocation_lng              738332 non-null  float64
 3   geolocation_city             738332 non-null  object 
 4   geolocation_state            738332 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 33.8+ MB


**Insight:**
- penghapusan duplikasi data sebanyak 261831 datum berhasil dilakukan.
- penghapusan dilakukan dari 1000163 baris menjadi 738332 baris.

## Exploratory Data Analysis (EDA)

#### orders_df

In [72]:
orders_df.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99441,99441,99441,99441
unique,99441,99441,8,98875,NaN,NaN,NaN,459
top,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-08-02 12:05:26,NaN,NaN,NaN,2017-12-20 00:00:00
freq,1,1,96478,3,NaN,NaN,NaN,522
mean,NaN,NaN,NaN,NaN,2017-12-31 19:07:21.164720896,2018-01-03 14:01:32.574989056,2018-01-12 20:22:20.863225344,NaN
min,NaN,NaN,NaN,NaN,2016-09-15 12:16:38,2016-10-06 06:06:04.500000,2016-10-11 13:46:32,NaN
25%,NaN,NaN,NaN,NaN,2017-09-12 21:55:13,2017-09-14 20:29:30,2017-09-22 19:22:41,NaN
50%,NaN,NaN,NaN,NaN,2018-01-19 11:54:30,2018-01-23 17:18:57,2018-02-01 12:03:58,NaN
75%,NaN,NaN,NaN,NaN,2018-05-04 23:50:21,2018-05-08 09:09:00,2018-05-15 16:26:01,NaN
max,NaN,NaN,NaN,NaN,2018-10-17 17:30:18,2018-10-19 13:09:42.500000,2018-10-26 15:33:22.500000,NaN


In [67]:
orders_df.groupby(by="order_status").order_id.nunique().sort_values(ascending=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: order_id, dtype: int64

In [68]:
orders_df.groupby(by="order_estimated_delivery_date").order_id.nunique().sort_values(ascending=False)

order_estimated_delivery_date
2017-12-20 00:00:00    522
2018-03-12 00:00:00    516
2018-03-13 00:00:00    513
2018-05-29 00:00:00    513
2018-02-14 00:00:00    507
                      ... 
2018-07-06 00:00:00      1
2018-07-10 00:00:00      1
2018-10-11 00:00:00      1
2018-10-30 00:00:00      1
2018-11-12 00:00:00      1
Name: order_id, Length: 459, dtype: int64

**penjelasan:**

untuk melihat berapa estimasi waktu yang diberikan dari sistem, kita perlu mengubah order_estimated_delivery_date menjadi "datetime", lalu menghitung selisih antara order_delivered_carrier_date dan order_estimated_delivery_date

In [75]:
orders_df["order_estimated_delivery_date"] = pd.to_datetime(orders_df["order_estimated_delivery_date"])

In [76]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  object        
 4   order_approved_at              99441 non-null  datetime64[ns]
 5   order_delivered_carrier_date   99441 non-null  datetime64[ns]
 6   order_delivered_customer_date  99441 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](4), object(4)
memory usage: 6.1+ MB


In [79]:
# melihat estimasi waktu pengiriman
# kemudian membandingkan dengan waktu pengiriman sebenarnya

estimated_time = orders_df["order_estimated_delivery_date"] - orders_df["order_delivered_carrier_date"]
estimated_time = estimated_time.apply(lambda x: x.total_seconds())
orders_df["estimated_time"] = round(estimated_time/86400)

delivery_time = orders_df["order_delivered_customer_date"] - orders_df["order_delivered_carrier_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)

In [80]:
orders_df.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,estimated_time,delivery_time
count,99441,99441,99441,99441,99441,99441,99441,99441,99441.000000,99441.000000
unique,99441,99441,8,98875,NaN,NaN,NaN,NaN,NaN,NaN
top,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-08-02 12:05:26,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,1,96478,3,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,2017-12-31 19:07:21.164720896,2018-01-03 14:01:32.574989056,2018-01-12 20:22:20.863225344,2018-01-24 03:08:37.730111232,20.390684,9.224384
min,NaN,NaN,NaN,NaN,2016-09-15 12:16:38,2016-10-06 06:06:04.500000,2016-10-11 13:46:32,2016-09-30 00:00:00,-147.000000,-16.000000
25%,NaN,NaN,NaN,NaN,2017-09-12 21:55:13,2017-09-14 20:29:30,2017-09-22 19:22:41,2017-10-03 00:00:00,15.000000,4.000000
50%,NaN,NaN,NaN,NaN,2018-01-19 11:54:30,2018-01-23 17:18:57,2018-02-01 12:03:58,2018-02-15 00:00:00,20.000000,7.000000
75%,NaN,NaN,NaN,NaN,2018-05-04 23:50:21,2018-05-08 09:09:00,2018-05-15 16:26:01,2018-05-25 00:00:00,26.000000,12.000000
max,NaN,NaN,NaN,NaN,2018-10-17 17:30:18,2018-10-19 13:09:42.500000,2018-10-26 15:33:22.500000,2018-11-12 00:00:00,192.000000,205.000000


In [87]:
orders_df[orders_df["estimated_time"] == -147][["order_id", "estimated_time", "delivery_time", "order_delivered_carrier_date", "order_estimated_delivery_date"]]


,order_id,estimated_time,delivery_time,order_delivered_carrier_date,order_estimated_delivery_date
47552,1612081119e8f23745698ad3367cc14b,-147.0,7.0,2017-04-13 10:57:02.500,2016-11-17


In [89]:
orders_df[orders_df["delivery_time"] == -16][["order_id", "estimated_time", "delivery_time", "order_delivered_carrier_date", "order_delivered_customer_date"]]


,order_id,estimated_time,delivery_time,order_delivered_carrier_date,order_delivered_customer_date
34939,c1e2bf2b7dd3309f2f5356c6b63968fa,12.0,-16.0,2017-03-02 17:34:26,2017-02-14 15:15:57


**penjelasan:**

  disini dapat dilihat bahwa terdapat estimated_time dan delivery_time yang bernilai **negatif**, hal ini dapat disebabkan oleh pengisian "order_delivered_carrier_date" atau "order_delivered_customer_date" yang kosong lalu diisi dengan median (pada tahap data cleaning).

  untuk mengatasinya, kita dapat mengubah order_delivered_carrier_date.

In [103]:
# mengisi missing value pada order_delivered_carrier_date
avg_delivery_time = (orders_df["order_delivered_customer_date"] - orders_df["order_delivered_carrier_date"]).median()
avg_estimated_time = (orders_df["order_estimated_delivery_date"] - orders_df["order_delivered_carrier_date"]).median()

In [104]:
# Update nilai order_delivered_carrier_date untuk order_id "1612081119e8f23745698ad3367cc14b"
orders_df.loc[orders_df["order_id"] == "1612081119e8f23745698ad3367cc14b", "order_delivered_carrier_date"] = (
    orders_df.loc[orders_df["order_id"] == "1612081119e8f23745698ad3367cc14b", "order_delivered_customer_date"] - pd.to_timedelta(avg_delivery_time)
)

In [106]:
orders_df.loc[orders_df["order_id"] == "c1e2bf2b7dd3309f2f5356c6b63968fa", "order_delivered_carrier_date"] = (
    orders_df.loc[orders_df["order_id"] == "c1e2bf2b7dd3309f2f5356c6b63968fa", "order_delivered_customer_date"] - avg_delivery_time
)

In [107]:
orders_df.loc[orders_df["order_id"] == "c1e2bf2b7dd3309f2f5356c6b63968fa", [
    "order_id",
    "delivery_time",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]]


,order_id,delivery_time,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
34939,c1e2bf2b7dd3309f2f5356c6b63968fa,-16.0,2017-02-07 12:52:17,2017-02-14 15:15:57,2017-03-15


In [105]:
orders_df.loc[orders_df["order_id"] == "1612081119e8f23745698ad3367cc14b", [
    "order_id",
    "estimated_time",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]]


,order_id,estimated_time,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
47552,1612081119e8f23745698ad3367cc14b,-147.0,2017-04-13 10:57:02.500,2017-04-20 13:20:42.500,2016-11-17


In [102]:
orders_df[orders_df["estimated_time"] == -147][["order_id", "estimated_time", "delivery_time", "order_delivered_carrier_date", "order_estimated_delivery_date"]]


,order_id,estimated_time,delivery_time,order_delivered_carrier_date,order_estimated_delivery_date
47552,1612081119e8f23745698ad3367cc14b,-147.0,7.0,2017-04-13 10:57:02.500,2016-11-17


#### order_items_df

In [60]:
order_items_df.describe(include="all")

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
count,112650,112650.000000,112650,112650,112650,112650.000000,112650.000000
unique,98666,NaN,32951,3095,93318,NaN,NaN
top,8272b63d03f5f79c56e9e4120aec44ef,NaN,aca2eb7d00ea1a7b8ebd4e68314663af,6560211a19b47992c3666cc44a7e94c0,2018-03-01 02:50:48,NaN,NaN
freq,21,NaN,527,2033,21,NaN,NaN
mean,NaN,1.197834,NaN,NaN,NaN,120.653739,19.990320
std,NaN,0.705124,NaN,NaN,NaN,183.633928,15.806405
min,NaN,1.000000,NaN,NaN,NaN,0.850000,0.000000
25%,NaN,1.000000,NaN,NaN,NaN,39.900000,13.080000
50%,NaN,1.000000,NaN,NaN,NaN,74.990000,16.260000
75%,NaN,1.000000,NaN,NaN,NaN,134.900000,21.150000


In [52]:
order_items_df["order_id"].value_counts()

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
428a2f660dc84138d969ccd69a0ab6d5    15
9ef13efd6949e4573a18964dd1bbe7f5    15
                                    ..
ffe73fc1d73e03fdb50e63903ddfe882     1
ffe8851012fcdaf83de7b595fd5154b3     1
ffea20c7630343a6cd9e09858c1295cd     1
ffea406a848c8afe4dec22bf6290ba00     1
fffe41c64501cc87c801fd61db3f6244     1
Name: count, Length: 98666, dtype: int64

In [64]:
order_items_df.groupby(by="product_id").agg({
    "order_id": "count",
    "price": ["max", "min", "mean", "std"]
}).sort_values(by=("order_id", "count"), ascending=False)


order_id  price                             
                                    count    max    min        mean       std
product_id                                                                   
aca2eb7d00ea1a7b8ebd4e68314663af      527  109.9   69.9   71.364137  3.670723
99a4788cb24856965c36a24e339b6058      488   89.9   74.0   88.167131  3.489328
422879e10f46682990de24d770e7f83d      484   59.9   49.0   54.911612  4.372601
389d119b48cf3043d311335e499d9c6b      392   59.9   49.0   54.695383  4.524637
368c6c730842d78016ad823897a372db      388   59.9   49.0   54.270103  4.484718
...                                   ...    ...    ...         ...       ...
00716e5593e8eea55dbe7a29b72a70bc        1   37.0   37.0   37.000000       NaN
00738f5a17ec4450e88915357b085c1e        1   23.9   23.9   23.900000       NaN
0081787b18cce4acc2d1adec4520d7ba        1  521.9  521.9  521.900000       NaN
0082684bb4a60a862baaf7a60a5845ed        1   39.9   39.9   39.900000       NaN
0085dddda27567189f1dfd1eda7c286e        1  179.8  179.8  179.800000       NaN

[32951 rows x 5 columns]

#### products_df

In [61]:
products_df.describe(include="all")

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32951,32951,32951.000000,32951.000000,32951.000000,32951.000000,32951.000000,32951.000000,32951.000000
unique,32951,74,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,106392145fca363410d287a815be6de4,cama_mesa_banho,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,48.523656,768.227945,2.166975,2276.376802,30.814725,16.937422,23.196534
std,NaN,NaN,10.156155,629.658469,1.728063,4281.926387,16.914005,13.637175,12.078707
min,NaN,NaN,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,NaN,NaN,42.000000,344.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,NaN,NaN,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,NaN,NaN,57.000000,961.000000,3.000000,1900.000000,38.000000,21.000000,30.000000


In [65]:
products_df.groupby(by="product_category_name").product_id.nunique().sort_values(ascending=False)

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
casa_conforto_2                     5
fashion_roupa_infanto_juvenil       5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: product_id, Length: 74, dtype: int64

In [70]:
products_df.groupby(by="product_category_name").agg({
    "product_id": "count",
    "product_weight_g": ["max", "min", "mean", "std"]
}).sort_values(by=("product_id", "count"), ascending=False)


product_id product_weight_g                      \
                                   count              max    min         mean   
product_category_name                                                           
cama_mesa_banho                     3029          40425.0    0.0  2456.405084   
esporte_lazer                       2867          30000.0   50.0  2024.670736   
moveis_decoracao                    2657          30000.0    2.0  3008.266842   
beleza_saude                        2444          30000.0   50.0  1434.793781   
utilidades_domesticas               2335          30000.0   50.0  3020.793576   
...                                  ...              ...    ...          ...   
casa_conforto_2                        5           3750.0  425.0  1467.600000   
fashion_roupa_infanto_juvenil          5            600.0  100.0   324.000000   
pc_gamer                               3           2839.0  650.0  1429.666667   
seguros_e_servicos                     2           1225.0  400.0   812.500000   
cds_dvds_musicais                      1            550.0  550.0   550.000000   

                                            
                                       std  
product_category_name                       
cama_mesa_banho                3732.667484  
esporte_lazer                  3973.126852  
moveis_decoracao               4713.454728  
beleza_saude                   3848.294144  
utilidades_domesticas          4737.014341  
...                                    ...  
casa_conforto_2                1382.339068  
fashion_roupa_infanto_juvenil   223.338308  
pc_gamer                       1222.820646  
seguros_e_servicos              583.363094  
cds_dvds_musicais                      NaN  

[74 rows x 5 columns]

#### product_category_name_translation_df

In [62]:
product_category_name_translation_df.describe(include="all")

,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1


In [71]:
product_category_name_translation_df

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor
...,...,...
66,flores,flowers
67,artes_e_artesanato,arts_and_craftmanship
68,fraldas_higiene,diapers_and_hygiene
69,fashion_roupa_infanto_juvenil,fashion_childrens_clothes


#### customers_df

In [63]:
customers_df.describe(include="all")

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
count,99441,99441,99441.000000,99441,99441
unique,99441,96096,NaN,4119,27
top,274fa6071e5e17fe303b9748641082c8,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,sao paulo,SP
freq,1,17,NaN,15540,41746
mean,NaN,NaN,35137.474583,NaN,NaN
std,NaN,NaN,29797.938996,NaN,NaN
min,NaN,NaN,1003.000000,NaN,NaN
25%,NaN,NaN,11347.000000,NaN,NaN
50%,NaN,NaN,24416.000000,NaN,NaN
75%,NaN,NaN,58900.000000,NaN,NaN


#### geolocation_df

In [64]:
geolocation_df.describe(include="all")

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
count,738332.000000,738332.000000,738332.000000,738332,738332
unique,NaN,NaN,NaN,8011,27
top,NaN,NaN,NaN,sao paulo,SP
freq,NaN,NaN,NaN,79927,285976
mean,38316.086800,-20.998353,-46.461098,NaN,NaN
std,30632.496675,5.892315,4.393705,NaN,NaN
min,1001.000000,-36.605374,-101.466766,NaN,NaN
25%,12600.000000,-23.603061,-48.867822,NaN,NaN
50%,29144.000000,-22.873588,-46.647278,NaN,NaN
75%,65950.000000,-19.923336,-43.836974,NaN,NaN


### Explore ...

**Insight:**
- xxx
- xxx

## Visualization & Explanatory Analysis

### Pertanyaan 1:

### Pertanyaan 2:

**Insight:**
- xxx
- xxx

## Analisis Lanjutan (Opsional)

## Conclusion

- Conclution pertanyaan 1
- Conclution pertanyaan 2